src/backend/chat/custom/custom.py

```diff
@@ -38,133 +41,188 @@ def chat(self, chat_request: CohereChatRequest, **kwargs: Any) -> Any:
-                for file_path in file_paths:
-                    retrievers.append(tool.implementation(file_path, **tool.kwargs))
-            elif tool.category != Category.FileLoader:
-                retrievers.append(tool.implementation(**tool.kwargs))
+        Args:
+            message (str): The message to be processed
+            chat_history (List[Dict[str, str]]): The chat history
+            tools (list[Tool]): The tools to be invoked
+            conversation_id (str): The conversation ID
+            deployment_model (BaseDeployment): The deployment model
+            kwargs (Any): The keyword arguments
 
-        return retrievers
+        Returns:
+            Any: The tool results or the chat response, and a boolean indicating if a direct answer was generated
 
-    def get_tool_results(
-        self, message: str, tools: list[Tool], model: BaseDeployment
-    ) -> list[dict[str, Any]]:
+        """
         tool_results = []
-        tools_to_use = model.invoke_tools(message, tools)
-
-        tool_calls = tools_to_use.tool_calls if tools_to_use.tool_calls else []
-        for tool_call in tool_calls:
-            tool = AVAILABLE_TOOLS.get(tool_call.name)
-            if not tool:
-                logging.warning(f"Couldn't find tool {tool_call.name}")
-                continue
 
-            outputs = tool.implementation().call(
-                parameters=tool_call.parameters,
+        # If the tool is Read_File or SearchFile, add the available files to the chat history
+        # so that the model knows what files are available
+        tool_names = [tool.name for tool in tools]
+        if ToolName.Read_File in tool_names or ToolName.Search_File in tool_names:
+            chat_history = self.add_files_to_chat_history(
+                chat_history,
+                conversation_id,
+                kwargs.get("session"),
+                kwargs.get("user_id"),
             )
 
-            # If the tool returns a list of outputs, append each output to the tool_results list
-            # Otherwise, append the single output to the tool_results list
-            outputs = outputs if isinstance(outputs, list) else [outputs]
-            for output in outputs:
-                tool_results.append({"call": tool_call, "outputs": [output]})
+        self.logger.info(f"Invoking tools: {tools}")
+        stream = deployment_model.invoke_tools(
+            message, tools, chat_history=chat_history
+        )
+
+        # Invoke tools can return a direct answer or a stream of events with the tool calls
+        # If the second event is a tool call, the tools are invoked, and the results are returned
+        # Otherwise, the chat response is returned as a direct answer
+        stream_start_event = next(stream)
+
+        if stream_start_event is None:
+            yield [], False
+
+        second_event = next(stream)
+
+        if second_event["event_type"] == StreamEvent.TOOL_CALLS_GENERATION:
```

Comment:
It could also be a tool calls partial generation, I think it is better to loop thorough the whole stream and when we see tool calls generation then call the code below

Reply:
can we get partial generation with single step?